### PR TITLE
Allow actions to opt into path mapping via execution requirements

### LINKED ISSUE #19526: Provide way to opt in arbitrary action types

### PR DESCRIPTION
**Pull request by fmeum**
This provides a simple and backwards compatible way for actions to experiment with path mapping:
* rules authors can add `supports-path-mapping` to the `execution_requirements` of a Starlark action
* end users can use tags and `--modify_execution_info` to opt individual targets or actions in and out of path mapping.

Also includes `StrippingPathMapperTest`, which wasn't included in the merge of #18098, with an additional test for a Starlark rule that opts in via the mechanisms described above.

Closes #19526
Work towards #6526

### CODE REVIEW

**Comment by aranguyen** [comment]
@fmeum I just realized the test wasn't included in the last merge. The patching tool failed and I had to manually resolve the differences and missed the test. Sorry about that. 

I saw a few different cls, could you please share here their priorities and the order I should be reviewing them in? THanks!

File: src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java
Comment (aranguyen) [inline src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:80]: does this method actually throw any exception?

File: src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java
Comment (aranguyen) [inline src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:151]: @fmeum Just thinking out loud here, I wonder if we should only allow the opt-in behavior for Bazel only because within google, we probably don't want users to flip path mapping on and off as it would cause forking (mapped and unmapped paths) and potentially harm caching? @gregestren WDYT?

**Comment by fmeum** [comment]
@aranguyen For Greg and you, only the current PR and then https://github.com/bazelbuild/bazel/pull/19723 should matter. https://github.com/bazelbuild/bazel/pull/19718, https://github.com/bazelbuild/bazel/pull/19719, and https://github.com/bazelbuild/bazel/pull/19721 can be reviewed by other folks, I just CCed you on them. After those, it's just end-to-end tests (https://github.com/bazelbuild/bazel/pull/18155).

File: src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:80]: It doesn't. For some reason my IDE thought it did.

File: src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:150]: Sure, I could see that make sense. I will leave that up to, happy to add this here.

Comment (aranguyen) [reply src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:150]: If doing so isn't too much work, can we please add the check then?

Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:150]: I thought I could use `isThisBazel`, but that is only available in tests. I don't see an easy and clean way to tell apart Bazel from Blaze. I could potentially look at the execpath of the output artifact of the action, but that seems hacky. Are you aware of a better way to implement the check?

Comment (aranguyen) [reply src/main/java/com/google/devtools/build/lib/analysis/actions/PathMappers.java:150]: I thought there was something similar to `JavaClassPathMode.BAZEL` but i don't think there is. I can check with others tomorrow. This shouldn't block this pr though. I will work on the import and work on the next one. Thanks!

**Comment by aranguyen** [comment]
> @aranguyen For Greg and you, only the current PR and then #19723 should matter. #19718, #19719, and #19721 can be reviewed by other folks, I just CCed you on them. After those, it's just end-to-end tests (#18155).

Thanks for clarifying. We do need all of these pr merge for path mappings to work in bazel right?

**Comment by fmeum** [comment]
> > @aranguyen For Greg and you, only the current PR and then #19723 should matter. #19718, #19719, and #19721 can be reviewed by other folks, I just CCed you on them. After those, it's just end-to-end tests (#18155).
> 
> Thanks for clarifying. We do need all of these pr merge for path mappings to work in bazel right?

Yes, that's correct. #19723 isn't required for just Java support though, only for good Starlark support.

**Comment by fmeum** [comment]
> Drive-by comment; apologies if this makes no sense. Now that we're adding an execution requirement, should we consider explicitly tagging the Java actions with it instead of making a decision based on the action mnemonic? Mnemonics are not unique and this opens us up to future cleanup work (i.e., someone reuses a mnemonic, gets path mapping behavior, and then we can't take it back because it's a breaking change).

This makes a lot of sense and is a very logical cleanup step, it's just something that a Googler will have to do as I can't test the change and probably don't have visibility into all affected rules.

File: src/test/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapperTest.java
Comment (aranguyen) [inline src/test/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapperTest.java:23]: @fmeum these tests don't work out very well internally. The paths are different `blaze-out` vs `bazel-out`. Instead of asserting for the exact paths, can we do regex matching instead? 

I also notice the order could be different as well. For example:

```
missing (6)
#1      : bazel-out/bin/java/com/google/test/B.java
#2      : bazel-out/bin/java/com/google/test/C.java
#3      : bazel-out/bin/java/com/google/test/liba-hjar.jar
#4      : bazel-out/bin/java/com/google/test/liba-hjar.jdeps
#5      : -XepOpt:foo:bar=bazel-out/bin/java/com/google/test/B.java
#6      : -XepOpt:baz=bazel-out/bin/java/com/google/test/C.java,bazel-out/bin/java/com/google/test/B.java

unexpected (6)
#1      : blaze-out/bin/java/com/google/test/liba-hjar.jar
#2      : blaze-out/bin/java/com/google/test/liba-hjar.jdeps
#3      : blaze-out/bin/java/com/google/test/B.java
#4      : blaze-out/bin/java/com/google/test/C.java
#5      : -XepOpt:foo:bar=blaze-out/bin/java/com/google/test/B.java
#6      : -XepOpt:baz=blaze-out/bin/java/com/google/test/C.java,blaze-out/bin/java/com/google/test/B.java
```

File: src/test/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapperTest.java
Comment (aranguyen) [inline src/test/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapperTest.java:82]: this has to be `scratch.file("defs/BUILD");` for internal tests to run. Would that be an issue?

File: src/test/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapperTest.java
Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/analysis/actions/StrippingPathMapperTest.java:23]: I'm not a big fan of regexes here as they may make our test assertions less precise. I am instead using the product name now.

I am not asserting on the order in the Java test, only in the Starlark test where it should be identical between Bazel and Blaze.

**Review by aranguyen — APPROVED** [review]
updated looks good after running internal tests

**Pull request closed without merging**
